service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+				BatchSize:     c.options.BatchSize(),
+			})
+		return err
+	}
+
+	err := backoff.Retry(op, persistenceOperationRetryPolicy, persistence.IsBackgroundTransientError)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return response.Tasks, len(response.NextPageToken) != 0, nil
+}
+
+func (c *crossClusterQueueProcessorBase) pollTasks() []task.Task {
+
+	var result []task.Task
+	for _, queueTask := range c.getTasks() {
+		crossClusterTask, ok := queueTask.(task.CrossClusterTask)
+		if !ok {
+			panic("received non cross cluster task")
+		}
+		if crossClusterTask.IsReadyForPoll() {
+			result = append(result, crossClusterTask)
+		}
+	}
+	return result
+}
+
+func (c *crossClusterQueueProcessorBase) getTask(key task.Key) (task.Task, error) {
+	for _, collection := range c.processingQueueCollections {
+		if task, err := collection.GetTask(key); err == nil {
+			return task, nil
+		}
+	}
+	return nil, errTaskNotFound
+}
+
+func (c *crossClusterQueueProcessorBase) getTasks() []task.Task {
+	var outstandingTask []task.Task
+	for _, collection := range c.processingQueueCollections {
+		outstandingTask = append(outstandingTask, collection.GetTasks()...)
+	}
+	return outstandingTask
+}
+
+func (c *crossClusterQueueProcessorBase) updateTask(
+	taskKey task.Key,
+	remoteResponse interface{},
+) error {
+
+	queueTask, err := c.getTask(taskKey)
+	if err != nil {
+		return err
+	}
+	crossClusterTask, ok := queueTask.(task.CrossClusterTask)
+	if ok {
+		if err := crossClusterTask.Update(remoteResponse); err != nil {
+			c.logger.Error("failed to update cross cluster task",
+				tag.TaskID(crossClusterTask.GetTaskID()),
+				tag.WorkflowDomainID(crossClusterTask.GetDomainID()),
+				tag.WorkflowID(crossClusterTask.GetWorkflowID()),
+				tag.WorkflowRunID(crossClusterTask.GetRunID()),
+				tag.Error(err))
+			return err
+		}
+	} else {
+		panic("Received non cross cluster task.")
+	}
+
+	// if the task fails to submit to main queue, it will use redispatch queue
+	_, err = c.submitTask(crossClusterTask)
+	return err
+}
+
+func (c *crossClusterQueueProcessorBase) addOutstandingTaskCountLocked(count int) {
+	c.outstandingTaskCount += count
+}
+
+// TODO: this will be called in cross cluster task Acked function.
+func (c *crossClusterQueueProcessorBase) removeOutstandingTaskCountLocked(count int) {
+	c.outstandingTaskCount -= count
+}
+
+func (c *crossClusterQueueProcessorBase) notifyAllQueueCollections() {
+	for _, queueCollection := range c.processingQueueCollections {
+		c.readyForProcess(queueCollection.Level())
+	}
+}
+
+func (c *crossClusterQueueProcessorBase) readyForProcess(level int) {
+	c.processingLock.Lock()
+	defer c.processingLock.Unlock()
+
+	if _, ok := c.backoffTimer[level]; ok {
+		// current level is being throttled
+		return
+	}
+
+	c.shouldProcess[level] = true
+
+	// trigger the actual processing
+	select {
+	case c.processCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *crossClusterQueueProcessorBase) resetBackoffTimer() {
```

Comment:
- I guess it also works if we just call the `setupBackoffTimer` for all queue levels? Whether the existing backoff timer is overwritten or not probably doesn't matter too much?
- With the current impl, I think you'll need to `Stop()` the exiting backoffTimer in the map if there's one?

Reply:
Updated. The stop() is already been called in processor stop()